### PR TITLE
Feature: Reusable Panel component

### PR DESCRIPTION
**Pull request by linux-nerd**
# Details

## Description
- [x] Create Panel, PanelHeader, PanelContent, PanelFooter components
- [x] Unit test cases
- [x] create story for illustration purpose
- [x] Add Panel without modal functionality

@crysfel I have added trigger props which will receive the Node. That node will be mounted and when that mounted node is clicked the modal will be displayed. I feel this is better approach as the developer does not have to take the responsibility of showing the popup when the node is clicked.

## Screenshots

screen shot 2018-10-10 at 2 36 35 pm
screen shot 2018-10-10 at 2 36 53 pm
screen shot 2018-10-11 at 4 33 36 pm

## Issue
#127

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,7 +39,7 @@
     "**/src/**/*.js": [
       "prettier --write",
       "eslint --fix",
-      "test:staged",
+      "yarn run test:staged",
```
Comment (nitish24p) [inline package.json]: i think you should change this to `npm` dont know if everyone is using yarn

File: src/components/panel/assets/panel.scss [hunk 2]
```diff
@@ -0,0 +1,37 @@
+@import '~theme/base/variables';
+
+.modal {
```
Comment (crysfel) [inline src/components/panel/assets/panel.scss]: It seems like basic styles for the panel (without rendering it in a modal) are missing.

I'm sorry for not being clear in the origin issue, but when the `modal` prop is equals to `false`, the panel should be able to render in place, like in a sidebar or anywhere else.

File: src/components/panel/assets/panel.scss [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+.modalMain {
+  position:fixed;
+  background: $color-white;
+  min-width: 40%;
+  height: auto;
+  top:50%;
+  left:50%;
+  transform: translate(-50%,-50%);
+  padding: 24px;
```
Comment (crysfel) [inline src/components/panel/assets/panel.scss]: Please add the `border-radius` as in the design

File: src/components/panel/assets/panelcontent.scss [hunk 4]
```diff
@@ -0,0 +1,7 @@
+@import '~theme/base/variables';
+
+.mainContent{
+  margin-bottom: 20px;
+  border: 2px solid $color-grey-subtle;
```
Comment (crysfel) [inline src/components/panel/assets/panelcontent.scss]: Please remove the `border`

File: src/components/panel/Panel.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  static Content = Content;
+  static Footer = Footer;
+
+  constructor(props) {
+    super(props);
+    // this.state = { open: this.props.open !== undefined ? this.props.open : Panel.defaultProps.open };
+    this.state = { ...Panel.defaultProps };
+
+    // event handlers hard bound to this
+    this.handleTriggerClick = this.handleTriggerClick.bind(this);
```
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/Panel.js]: @crysfel Any specific reason to choose arrow functions (class field syntax) over bind? React docs recommend both as best practice.

File: src/components/panel/Panel.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    });
+
+    const mainSectionClasses = cx({
+      modalMain: modal,
+      PanelMain: !modal,
+      [classes.body]: classes.body,
+    });
+
+    return [
+      trigger ? cloneElement(trigger, { onClick: this.handleTriggerClick, key: 'trigger' }) : null,
```
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/Panel.js]: Well, in that case don't use the trigger prop and instead handle the opening of the button from the component itself.

Something like this - 
```
class DummyComponent extends React.Component {
      constructor(props) {
        super(props);
        this.state = { open: false };
        this.closeModalHandler = e => this.setState({ open: false });
        this.triggerHandler = e => {
          alert('onTrigger handler called before opening the modal');
          this.setState({ open: true });
        };
      }

      render() {
        return (
          <React.Fragment>
            <Button text="Open Modal" onClick={this.triggerHandler} />
            <Panel
              closable={true}
              modal={true}
              open={this.state.open}
            >
              <Panel.Header>Hello World!</Panel.Header>
              <Panel.Content>
                <p>Trigger is not mandatory. If you have to do some processing before opening the modal/popup, then handle it seperately and then set open to `true`</p>
              </Panel.Content>
              <Panel.Footer>
                <Button text="Deny" onClick={this.closeModalHandler} />
                <Button text="Accept" onClick={this.closeModalHandler} />
              </Panel.Footer>
            </Panel>
          </React.Fragment>
        );
      }
    }

    return <DummyComponent />;
  }
```

I have added this in the story as well.

File: src/components/panel/Panel.stories.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      </Panel.Content>
+      <Panel.Footer>
+        <Button text="Deny" onClick={closeModal} />
+        <Button text="Accept" onClick={closeModal} />
+      </Panel.Footer>
+    </Panel>
+  );
+});
+
+storiesOf('Panel Component with external close control', module).add(
```
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/Panel.stories.js]: Even I like this approach. But I could not find beforeEach equivalent in RTL. I will check if this is possible and update the ut cases accordingly

[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/Panel.stories.js]: Can you explain basic panel precisely? :)

File: src/components/panel/PanelFooter.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  const classes = restProps.classes || {};
+
+  const footerClasses = cx({
+    footer: true,
+    [classes.footer]: classes.footer,
+  });
+
+  return (
+    <section className={footerClasses}>
+      <div className={styles.pullRight}>{children}</div>
```
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/PanelFooter.js]: Extra div is required to pull the content towards right hand side without using float.

File: src/components/panel/__tests__/Panel.test.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+  const triggerBtn = queryByText('Open Panel');
+  expect(triggerBtn).toBeInTheDocument();
+
+  fireEvent.click(triggerBtn);
+  const panel = await waitForElement(() => queryByText('A basic Panel'));
+
+  const closeIcon = container.querySelector('i');
+  fireEvent.click(closeIcon);
+  expect(handleClick).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/__tests__/Panel.test.js]: Event I am used to this approach. It makes me feel good about the total number of test cases :)
Actually I couldn't find any `beforeEach` equivalent in RTL, that's why I became lazy. I will check how to split it efficiently!

File: src/components/panel/assets/panel.scss (hunk 2 above)
Comment (linux-nerd) [reply src/components/panel/assets/panel.scss]: By only panel, do you mean just with backdrop and not positioned fixed?

File: package.json (hunk 1 above)
Comment (linux-nerd) [reply package.json]: I see package-lock.json file in the repo. I think changing yarn to npm will be better.

File: src/components/panel/assets/panel.scss (hunk 2 above)
Comment (tureey) [reply src/components/panel/assets/panel.scss]: I think so

File: src/components/panel/__tests__/Panel.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/__tests__/Panel.test.js]: @crysfel I was split the test in multiple tests.

File: src/components/panel/assets/panel.scss (hunk 2 above)
Comment (linux-nerd) [reply src/components/panel/assets/panel.scss]: @crysfel, @tureey I have added styles for non modal popup

File: src/components/panel/PanelFooter.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (linux-nerd) [reply src/components/panel/PanelFooter.js]: @crysfel Is it the responsibility of the component to pull the content to right. This looks fine for actions, but not for some text. What do you think?

File: src/components/panel/Panel.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Venablena) [reply src/components/panel/Panel.js]: Arrow functions are more succinct (yes!!!), I'm always in favor of reducing boilerplate and complexity. There are plenty of differing opinions on this (too) out there, though :).

File: src/components/panel/PanelFooter.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (crysfel) [reply src/components/panel/PanelFooter.js]: maybe we can define a property to set the alignment, by default nothing is applied. what about that?

File: src/components/panel/__tests__/Panel.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (crysfel) [reply src/components/panel/__tests__/Panel.test.js]: >  It makes me feel good about the total number of test cases :)
same here! 😂

Well... I don't see any reason not to use:

```
describe('components/Panel', () => {
  it('...', ()=> ...);

  it('...', ()=> ...);
});
```

You will then have your `beforeEach`.

File: src/components/panel/assets/panel.scss (hunk 2 above)
Comment (crysfel) [reply src/components/panel/assets/panel.scss]: yeap!

File: src/components/panel/Panel.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tureey) [reply src/components/panel/Panel.js]: I agree with @crysfel

File: src/components/panel/Panel.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+import classNames from 'classnames/bind';
+import styles from './assets/panel.scss';
+import Header from './PanelHeader';
+import Content from './PanelContent';
+import Footer from './PanelFooter';
+import { noop } from 'utils/noop';
+
+const cx = classNames.bind(styles);
+
+/**
```
Comment (tureey) [inline src/components/panel/Panel.js]: Would be nice a hook to delete all the comments for production code

**Pull request merged**